compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -4684,6 +4691,68 @@ private List<BAnnotationAttachmentSymbol> getAnnotationAttachmentSymbols(
         return annotationAttachmentSymbols;
     }
 
+    private void validateTypesOfOverriddenFields(BLangStructureTypeNode structureTypeNode) {
+        validateTypesOfOverriddenFields(structureTypeNode.getBType(), structureTypeNode.fields,
+                                        structureTypeNode.typeRefs);
+    }
+
+    private void validateTypesOfOverriddenFields(BType type, List<BLangSimpleVariable> fields,
+                                                 List<BLangType> includedTypeNodes) {
+        if (type == symTable.semanticError) {
+            return;
+        }
+
+        LinkedHashMap<String, BField> fieldsOfIncludingType = ((BStructureType) type).fields;
+        Map<String, Location> explicitlySpecifiedFieldLocations = getFieldLocations(fields);
+
+        for (BLangType includedTypeNode : includedTypeNodes) {
+            BType includedType = Types.getReferredType(includedTypeNode.getBType());
+
+            int includedTypeTag = includedType.tag;
+            if (includedTypeTag != TypeTags.RECORD && includedTypeTag != TypeTags.OBJECT) {
+                continue;
+            }
+
+            BStructureType includedStructureType = (BStructureType) includedType;
+
+            for (Map.Entry<String, BField> includedFieldEntry : includedStructureType.fields.entrySet()) {
+                String fieldName = includedFieldEntry.getKey();
+
+                if (!explicitlySpecifiedFieldLocations.containsKey(fieldName) ||
+                        // Happens when the type cannot be resolved.
+                        !fieldsOfIncludingType.containsKey(fieldName)) {
+                    continue;
+                }
+
+                BType fieldTypeInIncludingType = fieldsOfIncludingType.get(fieldName).type;
+                if (fieldTypeInIncludingType == symTable.semanticError) {
+                    continue;
+                }
+
+                BType fieldTypeInIncludedType = includedFieldEntry.getValue().type;
+                if (fieldTypeInIncludedType == symTable.semanticError) {
+                    continue;
+                }
```

Comment:
Scenario 1:
```ballerina
type Bar record {|
    record {string x;} body;
|};


public type Baz record {|
    *Bar;
    Mat body;
|};
```
Scenario 2:
```ballerina
type Bar record {|
    record {string x;} body;
|};


public type Baz record {|
    *Bar;
    record {Mat x;} body;
|};

```
for the first scenario, we are getting an error as
```
ERROR [test3.bal:(13:5,13:8)] unknown type 'Mat'

```
For the second scenario, we are getting an error as
```
ERROR [test3.bal:(13:5,13:26)] included field 'body' of type 'record {| string x; anydata...; |}' cannot be overridden by a field of type 'record {| anydata...; |}': expected a subtype of 'record {| string x; anydata...; |}'
ERROR [test3.bal:(13:13,13:16)] unknown type 'Val'

```
this is something like two behaviours for the same scenario. Isn't it?
anyway for the second scenario error message is incorrect.

scenario 3
```
type Bar record {|
    record {} body;
|};


public type Baz record {|
    *Bar;
    record {Mat x;} body;
|};

```
When considering scenario 3 we get only the 
```
ERROR [test3.bal:(13:13,13:16)] unknown type 'Mat'

```

Reply:
The behaviour is due to how we resolve `record {Mat x;}` when `Mat` is undefined, right? We seem to be resolving it as `record {}` since we don't know the type of `x` to define it as a field, which IMO is acceptable.

This isn't specific to the changes introduced in this PR either. The errors are similar for the following, which does pretty much the same checks.

```ballerina
function fn(record {Mat x;} a) { // unknown type 'Mat'
    record {string x;} _ = a; // error: incompatible types: expected 'record {| string x; anydata...; |}', found 'record {| anydata...; |}'
    record {} _ = a; // no error
}
```

Reply:
okay. 
But here we check whether the type is defined or not. As an example
```ballerina
type Bar record {|
    record {string x;} body;
|};


public type Baz record {|
    *Bar;
    Mat body; // we check whether `Mat` defined or not.
|};
```
if it is an anon record type we didn't check types of fields of the anon type are defined or not.
So do we need these checks here?
```
if (fieldTypeInIncludingType == symTable.semanticError) {
    continue;
}

if (fieldTypeInIncludedType == symTable.semanticError) {
    continue;
}
```

Reply:
The type being undefined is different from a field of a type being undefined. Without these checks, I believe there'll be errors like

```cmd
included field 'body' of type 'anydata' cannot be overridden by a field of type 'other': expected a subtype of 'anydata'
``` 

which don't add any value.